File: io.openems.common/src/io/openems/common/jsonrpc/notification/SystemLogNotification.java
package io.openems.common.jsonrpc.notification;

import org.ops4j.pax.logging.spi.PaxLoggingEvent;

import com.google.gson.JsonObject;

import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
import io.openems.common.jsonrpc.base.JsonrpcNotification;
import io.openems.common.types.SystemLog;
import io.openems.common.utils.JsonUtils;

/**
 * Represents a JSON-RPC Notification for sending the current system log.
 *
 * <pre>
 * {
 *   "jsonrpc": "2.0",
 *   "method": "systemLog",
 *   "params": {
 *     "line": {@link SystemLog#toJson()}
 *   }
 * }
 * </pre>
 */
public class SystemLogNotification extends JsonrpcNotification {

	public static final String METHOD = "systemLog";

	private final SystemLog line;

	/**
	 * Parses a {@link JsonrpcNotification} to a {@link SystemLogNotification}.
	 *
	 * @param n the {@link JsonrpcNotification}
	 * @return the {@link SystemLogNotification}
	 * @throws OpenemsNamedException on error
	 */
	public static SystemLogNotification from(JsonrpcNotification n) throws OpenemsNamedException {
		var j = n.getParams();
		var line = SystemLog.fromJsonObject(JsonUtils.getAsJsonObject(j, "line"));
		return new SystemLogNotification(line);
	}

	/**
	 * Creates a {@link SystemLogNotification} from a {@link PaxLoggingEvent}.
	 *
	 * @param event the {@link PaxLoggingEvent}
	 * @return the {@link SystemLogNotification}
	 */
	public static SystemLogNotification fromPaxLoggingEvent(PaxLoggingEvent event) {
		return new SystemLogNotification(SystemLog.fromPaxLoggingEvent(event));
	}

	public SystemLogNotification(SystemLog line) {
		super(SystemLogNotification.METHOD);
		this.line = line;
	}

	@Override
	public JsonObject getParams() {
		return JsonUtils.buildJsonObject() //
				.add("line", this.line.toJson()) //
				.build();
	}

}

File: ui/src/app/shared/jsonrpc/notification/systemLogNotification.ts
import { JsonrpcNotification } from "../base";
import { SystemLog } from "../../type/systemlog";

/**
 * Represents a JSON-RPC Notification for sending the current system log.
 *
 * <pre>
 * {
 *   "jsonrpc": "2.0",
 *   "method": "systemLog",
 *   "params": {
 *     "line": {
 *       "time": string, // in the form '2011-12-03T10:15:30+01:00'
 *       "level": 'ERROR' | 'WARN' | 'INFO',
 *       "source": string,
 *       "message": string
 *     }
 *   }
 * }
 * </pre>
 */
export class SystemLogNotification extends JsonrpcNotification {

    public static readonly METHOD: string = "systemLog";

    public constructor(
        public override readonly params: {
            line: SystemLog
        },
    ) {
        super(SystemLogNotification.METHOD, params);
    }

}
